source-oracle/backfill.go

```diff
@@ -0,0 +1,383 @@
+
+	// Execute the backfill query to fetch rows from the database
+	logEntry.WithFields(logrus.Fields{"query": query, "args": args}).Debug("executing query")
+	rows, err := db.conn.QueryContext(ctx, query, args...)
+	if err != nil {
+		return false, fmt.Errorf("unable to execute query %q: %w", query, err)
+	}
+	defer rows.Close()
+
+	// Process the results into `changeEvent` structs and return them
+	cols, err := rows.Columns()
+	if err != nil {
+		return false, fmt.Errorf("rows.Columns: %w", err)
+	}
+	var resultRows int // Count of rows received within the current backfill chunk
+	var rowOffset = state.BackfilledCount
+	var prevRowID string // Used when processing keyless backfill results to sanity-check ordering
+	logEntry.Debug("translating query rows to change events")
+	for rows.Next() {
+		fields, err := scanToMap(rows, cols, columnTypes)
+		if err != nil {
+			return false, err
+		}
+		var rowid = fields["ROWID"].(string)
+		delete(fields, "ROWID")
+
+		var rowKey []byte
+		if state.Mode == sqlcapture.TableModeKeylessBackfill {
+			rowKey = []byte(rowid)
+
+			// Sanity check that rows are returned in ascending RowID order within a given backfill chunk
+			// Note that base64 strings are not lexicographically ordered and have a different ordering, hence
+			// the custom function
+			if result, err := compareBase64(rowid, prevRowID); err != nil {
+				return false, err
+			} else if result == -1 {
+				return false, fmt.Errorf("internal error: ROWID ordering sanity check failed: %v <= %v", rowid, prevRowID)
+			}
+			prevRowID = rowid
+		} else {
+			rowKey, err = sqlcapture.EncodeRowKey(keyColumns, fields, columnTypes, encodeKeyFDB)
+			if err != nil {
+				return false, fmt.Errorf("error encoding row key for %q: %w", streamID, err)
+			}
+		}
+
+		if err := translateRecordFields(info, fields); err != nil {
+			return false, fmt.Errorf("error backfilling table %q: %w", table, err)
+		}
+
+		var event = &sqlcapture.ChangeEvent{
+			Operation: sqlcapture.InsertOp,
+			RowKey:    rowKey,
+			Source: &oracleSource{
+				SourceCommon: sqlcapture.SourceCommon{
+					Millis:   0, // Not known.
+					Schema:   schema,
+					Snapshot: true,
+					Table:    table,
+				},
+				RowID: rowid,
+			},
+			Before: nil,
+			After:  fields,
+		}
+		if err := callback(event); err != nil {
+			return false, fmt.Errorf("error processing change event: %w", err)
+		}
+		resultRows++
+		rowOffset++
+	}
+
+	if err := rows.Err(); err != nil {
+		return false, err
+	}
+
+	var backfillComplete = resultRows < db.config.Advanced.BackfillChunkSize
+	return backfillComplete, nil
+}
+
+func scanToMap(rows *sql.Rows, cols []string, colTypes map[string]oracleColumnType) (map[string]any, error) {
+	var fieldsArr = make([]any, len(cols))
+	var fieldsPtr = make([]any, len(cols))
+	for idx, col := range cols {
+		if col == "ROWID" {
+			var rowid = ""
+			fieldsArr[idx] = &rowid
+		} else {
+			fieldsArr[idx] = reflect.Zero(colTypes[col].t).Interface()
```

Comment:
Does the oracle driver require the placeholder values to have the appropriate concrete type for the interfaces that will receive their values? So they can't be just an `interface{}`? Or get them via `rows.ColumnTypes`?

Reply:
Huh, I don't know why I thought it is necessary to have the concrete type for each column, I switched to using `new(any)` to simplify the code (and improve performance, reflect is time-consuming)